README
Thread::Cancel version 1.09
===========================

This module adds the capability to cancel (i.e., kill) a thread.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires Perl 5.8.0 or later.

This module requires these other modules:

    threads version 1.39 or later
    Test::More 0.50 or later (for installation)

COPYRIGHT AND LICENCE

Copyright 2006 - 2008 Jerry D. Hedden <jdhedden AT cpan DOT org>

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

# EOF


Changes
Revision history for Perl extension Thread::Cancel.

-
	-

1.13 Fri Dec 24 17:59:17 2010
	- POD update

1.11 Fri Apr 16 13:45:44 2010
	- Fix tests following recent changes to blead

1.09 Thu Jun 12 13:44:17 2008
	- Fix SKIPs in tests

1.08 Tue Jun 10 16:56:17 2008
	- End all tests with exit(0)

1.07 Fri Feb 22 22:12:47 2008
	- Allow installation on non-threaded Perls

1.06 Wed Feb 20 17:29:39 2008
	- Build/test updates

1.05 Sat Oct 14 09:17:09 EST 2006
	- Check for $Config{useithreads}

1.04 Thu Oct 12 11:23:09 EDT 2006
	- Added example and POD tests

1.03 Wed Aug 30 14:25:35 EDT 2006
	- Support 5.8.0

1.02 Tue Aug  1 14:05:41 EDT 2006
	- Fixes to tests

1.01 Tue Jul 11 14:38:35 EDT 2006
	- Initial release



examples/cancel.pl
#!/usr/bin/perl

use strict;
use warnings;

use threads;
use threads::shared;
use Thread::Cancel;
use Thread::Queue;

### Global Variables ###

# Maximum working threads
my $MAX_THREADS = 10;

# Maximum thread working time
my $TIMEOUT = 10;

# Flag to inform all threads that application is terminating
my $TERM :shared = 0;


### Signal Handling ###

# Gracefully terminate application on ^C
# or command line 'kill'
$SIG{'INT'} = $SIG{'TERM'} =
    sub {
        print(">>> Terminating <<<\n");
        $TERM = 1;
    };


### Main Processing Section ###
MAIN:
{
    # Start timer thread
    my $queue = Thread::Queue->new();
    threads->create('timer', $queue)->detach();

    # Manage the thread pool until signalled to terminate
    while (! $TERM) {
        # Keep max threads running
        for (my $needed = $MAX_THREADS - threads->list();
             $needed && ! $TERM;
             $needed--)
        {
            # New thread
            threads->new('worker', $queue, $TIMEOUT);
        }

        # Wait for any threads to finish
        sleep(1);
    }

    ### CLEANING UP ###

    # Wait for max timeout for threads to finish
    while ((threads->list() > 0) && $TIMEOUT--) {
        sleep(1);
    }

    # Detach and cancel any remaining threads
    $_->cancel() foreach (threads->list());
    sleep(1);
}

print("Done\n");
exit(0);


### Thread Entry Point Subroutines ###

# A worker thread
sub worker
{
    my ($queue, $timeout) = @_;

    ### INITIALIZE ###

    # My thread ID
    my $tid = threads->tid();
    printf("Working -> %3d\n", $tid);

    # Register with timer thread
    $queue->enqueue($tid, $timeout);


    ### WORK ###

    # Do some work while monitoring $TERM
    my $sleep = 5 + int(rand(10));
    while (($sleep > 0) && ! $TERM) {
        $sleep -= sleep($sleep);
    }


    ### DONE ###

    # Unregister with timer thread
    $queue->enqueue($tid, undef);

    # Tell user we're done
    printf("           %3d <- Finished\n", $tid);

    # Detach and terminate
    threads->detach() if ! threads->is_detached();
    threads->exit();
}


# The timer thread that monitors other threads for timeout
sub timer
{
    my $queue = shift;   # The registration queue
    my %timers;          # Contains threads and timeouts

    # Loop until told to quit
    while (! $TERM) {
        # Check queue
        while (my $tid = $queue->dequeue_nb()) {
            if (! ($timers{$tid}{'timeout'} = $queue->dequeue()) ||
                ! ($timers{$tid}{'thread'}  = threads->object($tid)))
            {
                # No timeout - unregister thread
                delete($timers{$tid});
            }
        }

        # Cancel timed out threads
        foreach my $tid (keys(%timers)) {
            if (--$timers{$tid}{'timeout'} < 0) {
                $timers{$tid}{'thread'}->cancel;
                printf("           %3d <- Cancelled\n", $tid);
                delete($timers{$tid});
            }
        }

        # Tick tock
        sleep(1);
    }
}

__END__

=head1 NAME

cancel.pl - Simple 'threads' example

=head1 DESCRIPTION

A simplistic example illustrating the following:

=over

=item * Management of a pool of threads

=item * Communication between threads using queues

=item * Timing out and cancelling threads

=item * Interrupting a threaded program

=item * Cleaning up threads before terminating

=back

=head1 SEE ALSO

L<threads>, L<threads::shared>, and L<Thread::Queue>

=head1 AUTHOR

Jerry D. Hedden, S<E<lt>jdhedden AT cpan DOT orgE<gt>>

=head1 COPYRIGHT AND LICENSE

Copyright 2006 - 2007 Jerry D. Hedden. All rights reserved.

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Thread/Cancel.pm
package Thread::Cancel; {

use strict;
use warnings;

our $VERSION = '1.13';

use threads 1.39;

my $SIGNAL = 'KILL';    # Default cancellation signal

sub import
{
    my $class = shift;   # Not used

    # Set the signal for cancel operations
    while (my $sig = shift) {
        $SIGNAL = $sig;
    }
    $SIGNAL =~ s/^SIG//;

    # Set up the cancel signal handler
    $SIG{$SIGNAL} = sub { threads->exit(); };
}


sub threads::cancel
{
    my ($class, @threads) = @_;

    if ($class eq 'threads') {
        if (@threads) {
            # Cancel specified list of threads
            @threads = grep { $_ }
                       map  { (ref($_) eq 'threads')
                                    ? $_
                                    : threads->object($_) }
                            @threads;
        } else {
            # Cancel all threads
            push(@threads, threads->list());
        }
    } else {
        # Cancel a single thread
        push(@threads, $class);
    }

    # Cancel threads
    my $resumable = threads->can('resume');
    foreach my $thr (@threads) {
        $thr->detach() if (! $thr->is_detached());
        $thr->kill($SIGNAL);
        if ($resumable) {
            $thr->resume() if ($thr->is_suspended());
        }
    }
}

}

1;

__END__

=head1 NAME

Thread::Cancel - Cancel (i.e., kill) threads

=head1 VERSION

This document describes Thread::Cancel version 1.13

=head1 SYNOPSIS

    use Thread::Cancel 'SIGUSR1';      # Set the cancellation signal
    use Thread::Cancel;                #  Defaults to 'KILL'

    $thr->cancel();                    # Cancel a thread
    threads->cancel();                 # Cancel all non-detached threads
    threads->cancel($thr, $tid, ...);  # Cancel multiple threads using
                                        #   objects or TIDs

=head1 DESCRIPTION

This module adds cancellation capabilities for threads.  Cancelled threads are
terminated using C<threads-E<gt>exit()>.  The thread is then detached, and
hence automatically cleaned up.

Threads that are suspended using L<Thread::Suspend> do not need to be
I<resumed> in order to be cancelled.

It is possible for a thread to cancel itself.

=head2 Declaration

This module must be imported prior to any threads being created.

Cancellation is accomplished via a signal handler which is used by all threads
on which cancel operations are performed.  The signal for this operation can
be specified when this module is declared, and defaults to C<SIGKILL>.
Consequently, the application and its threads must not specify some other
handler for use with the cancel signal.

=over

=item use Thread::Cancel;

Declares this module, and defaults to using C<SIGKILL> for cancel operations.

=item use Thread::Cancel 'SIGUSR1';

=item use Thread::Cancel 'Signal' => 11;

Declares this module, and uses the specified signal for cancel operations.
Signals may be specified by the same names or (positive) numbers as supported
by L<kill()|perlfunc/"kill SIGNAL, LIST">.

=back

=head2 Methods

=over

=item $thr->cancel()

Cancels the threads.

=item threads->cancel()

Cancels all non-detached threads.  This offers a clean way to exit a threaded
application:

    # Terminate all threads and exit
    threads->cancel();
    exit(0);

=item threads->cancel($thr, $tid, ...)

Cancels the threads specified by their objects or TIDs (for non-detached
threads).

=back

=head1 CAVEATS

Subject to the limitations of L<threads/"THREAD SIGNALLING">.

Cancelled threads are automatically detached, so do not try to C<-E<gt>join()>
or C<-E<gt>detach()> a cancelled thread.

Detached threads can only be cancelled using their I<threads> object:

    $thr->detach();
    $thr->cancel();
    # or
    threads->cancel($thr);

Threads that have finished execution are, for the most part, ignored by this
module.

=head1 REQUIREMENTS

Perl 5.8.0 or later

L<threads> 1.39 or later

L<Test::More> 0.50 or later (for installation)

=head1 SEE ALSO

Thread::Cancel on MetaCPAN:
L<https://metacpan.org/release/Thread-Cancel>

Code repository:
L<https://github.com/jdhedden/Thread-Cancel>

L<threads>, L<threads::shared>

Sample code in the I<examples> directory of this distribution on CPAN.

=head1 AUTHOR

Jerry D. Hedden, S<E<lt>jdhedden AT cpan DOT orgE<gt>>

=head1 COPYRIGHT AND LICENSE

Copyright 2006 - 2009 Jerry D. Hedden. All rights reserved.

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


Makefile.PL
# Module makefile for Thread::Cancel (using ExtUtils::MakeMaker)

use strict;
use warnings;

use ExtUtils::MakeMaker;

# Construct make file
WriteMakefile(
    'NAME'          => 'Thread::Cancel',
    'AUTHOR'        => 'Jerry D. Hedden <jdhedden AT cpan DOT org>',
    'VERSION_FROM'  => 'lib/Thread/Cancel.pm',
    'ABSTRACT_FROM' => 'lib/Thread/Cancel.pm',
    'PREREQ_PM'     => {
        'threads'         => 1.39,
        'threads::shared' => 0,
        'Test::More'      => 0.50,
    },

    ((ExtUtils::MakeMaker->VERSION() lt '6.25') ?
        ('PL_FILES' => { })            : ()),
    ((ExtUtils::MakeMaker->VERSION() gt '6.30') ?
        ('LICENSE'  => 'perl')         : ()),
);

# Additional 'make' targets
sub MY::postamble
{
    return <<'_EXTRAS_';
fixfiles:
	@dos2unix `cat MANIFEST`
	@$(CHMOD) 644 `cat MANIFEST`
	@$(CHMOD) 755 examples/*.pl
_EXTRAS_
}

# EOF
